// nodemon/lib/dist/spawn.dev.js
"use strict";

var utils = require('./utils');

var merge = utils.merge;
var bus = utils.bus;

var spawn = require('child_process').spawn;

module.exports = function spawnCommand(command, config, eventArgs) {
  var stdio = ['pipe', 'pipe', 'pipe'];

  if (config.options.stdout) {
    stdio = ['pipe', process.stdout, process.stderr];
  }

  var sh = 'sh';
  var shFlag = '-c';

  if (utils.isWindows) {
    sh = 'cmd';
    shFlag = '/c';
  }

  if (!Array.isArray(command)) {
    command = [command];
  }

  var args = command.join(' ');
  var env = merge(process.env, {
    FILENAME: eventArgs[0]
  });
  var child = spawn(sh, [shFlag, args], {
    env: merge(config.options.execOptions.env, env),
    stdio: stdio
  });

  if (config.required) {
    var emit = {
      stdout: function stdout(data) {
        bus.emit('stdout', data);
      },
      stderr: function stderr(data) {
        bus.emit('stderr', data);
      }
    }; // now work out what to bind to...

    if (config.options.stdout) {
      child.on('stdout', emit.stdout).on('stderr', emit.stderr);
    } else {
      child.stdout.on('data', emit.stdout);
      child.stderr.on('data', emit.stderr);
      bus.stdout = child.stdout;
      bus.stderr = child.stderr;
    }
  }
};
